World edit commands for Cubyz

Argmaster opened the pull request:
This pull request introduces some of the commands available in very popular minecraft mod called WorldEdit. 
List of WorldEdit commands can be found [here](https://minecraft-worldedit.fandom.com/wiki/Worldedit_Commands)
For best user experience I decided to take approach of implementing those commands as close to original as possible.

## CHANGELOG:

- [x] `/pos1`
- [x] `/pos2`
- [x] `/copy`
- [x] `/paste`
- [x] `/blueprint save`
- [x]  `/blueprint load`
- [x] `/blueprint delete`
- [x] `/blueprint list`
- [x] binary storage format
- [x] wireframe outline of selected region

Even if Cubyz engine is not ready for large amounts of block updates I think that players don't really care and will be happy to have those features despite potential lag spikes.

I do not plan adding brushes in this PR, it will be waaay too much code at once. For that reason `rotate` and `flip` are also considered optional, maybe they can be added in a basic version where they just move blocks around but don't update block data to eg. flip stairs. Pattern expressions are also outside of the scope of this PR, for now commands that take patterns will be only able to take one string block ID.

Related to: #1214

Argmaster commented:
Oh and also this PR adds schematic storage format as part of `//blueprint save`

Argmaster commented:
> If possible, can you add a wand item? It doesn't need a fancy texture

Sure, but in separate PR - we are already near 500 lines - I think that I should also leave rest of the features listed (maybe except schematic loading) for separate PR. @IntegratedQuantum won't be happy about reviewing my spagetti 🥲

OneAvargeCoder193 commented:
I'm pretty sure that the way pos1 and pos2 are currently stored, if two people are trying to work with worldedit at once on a server, they will override each other's positions.

Argmaster commented:
> At a glance I already see a few issues:
> 
> * Storing blueprints as zon is just about the worst thing you could do for file size, please store binary data + at least deflate compression.

I agree, but this is a quick way to do it and I want to support a human readable format for sake of being able to sit and scribble something in notepad if needed. As you already said, Cubyz is not ready for large amounts of block updates, so let's maybe avoid premature optimalization for sake of having features, what do you think?

Regardless of what I just said, I agree that this should not be a primary storage format, I can and will work on a binary format, but I don't want to do it in this PR, IMO this is beyond providing simple WorldEdit functionality, so it is beyond of my goals right now.

On top of that I would like to also at least partially support imports from minecraft schematic format to a reasonable extent with some limited translation table, but that is a material for yet another PR.

> * Shouldn't blueprints also save the blockdata, and not just the block type? 

I think they are doing that, the palette stores only block type, but block array stores both, unless there is a bug in the code.

> Also the use of a palette seems weird, looks like it would strictly makes memory usage and performance worse here.

You have to translate from block palette that was present in world you exported from to block palette of world you are importing into, at least because they could have been created on different versions of the game, so there must be a palette storage. Since most of the time you will not be using whole palette, there is also no point in storing whole palette, which can have up to 65k entries. Therefore palette is created with blocks that are necessary and reassigning IDs ensures there are no gaps in ID list.

For rotations and flips you will have to translate it back to world palette regardless of what palette you use for storage, as rotations and flips must also work on schematics you load from drive.

> * Saving blueprints on the server is probably not the right way to go about it.

This is what WorldEdit does, sure, we can do something different but then you have to transfer schematic to client and that is more complicated for limited benefit. Users of those commands are expected to have admin rights, thus also access to the server.

> * Overwriting the command id should not be needed, you can just use zig's `@""` syntax to create any name you want as an identifier.

That's great, I didn't know that!

> * Please don't introduce a new command prefix `//` without any good reason. There should be only one way to write commands everything else is just needlessly confusing.

I guess this is it for sticking with WorldEdit design. I know, It's Cubyz, not minecraft, but being unique is not always an advantage, especially from user experience it is usually better to follow standards that were established before.

OneAvargeCoder193 commented on src/server/command/worldedit/pos1.zig:
```diff
@@ -0,0 +1,23 @@
+pub const description = "Select the player's position as position 1.";
+pub const usage = "/pos1";
+
+pub var pos: Vec3i = .{0, 0, 0};
```
I don't think you should store the worldedit positions in the command folder, as multiple people can override each others positions. It would be best to store this data somewhere in the user struct I think, and then use ```source.pos1``` and ```source.pos2``` in the worldedit commands

OneAvargeCoder193 commented on src/server/command/worldedit/pos2.zig:
```diff
@@ -0,0 +1,23 @@
+pub const description = "Select the player's position as position 2.";
+pub const usage = "/pos2";
+
+pub var pos: Vec3i = .{0, 0, 0};
```
Same here.

OneAvargeCoder193 reviewed (changes requested):
I think there are a couple issues with the way that you store the selected blocks

IntegratedQuantum commented:
> I agree, but this is a quick way to do it and I want to support a human readable format for sake of being able to sit and scribble something in notepad if needed. As you already said, Cubyz is not ready for large amounts of block updates, so let's maybe avoid premature optimalization for sake of having features, what do you think?

The quick way it is going to screw us over in the long run. And it isn't even that much quicker than the alternative, instead of doing zon.get, you just read stuff at an offset. And `utils.Compression` makes it easy to compress the file data.
A binary format will be necessary at some point, and if we don't add it now, then we'll need to add conversion code.
And in comparison a binary format can be easily made future proof by adding a version number to the header.
And seriously, no one is going to scribble their structures in a text editor, when they could just build it in the game.

Also Cubyz may not be ready for large block updates, but if it's there, then I'm sure that @OneAvargeCoder193 will be tempted to use it for structure generation.

> On top of that I would like to also at least partially support imports from minecraft schematic format

That should be a third-party conversion tool.

> I think they are doing that, the palette stores only block type, but block array stores both, unless there is a bug in the code.

If it does that, then please use the `Block` struct for it for better readability, right now it's just a `u32`.

> You have to translate from block palette that was present in world you exported from to block palette of world you are importing into, at least because they could have been created on different versions of the game

Ah, if it's only for storing the file, then it should only be used once when loading it to translate the block ids. Doing this translation every time a structure is pasted, is just wasteful.

> This is what WorldEdit does, sure, we can do something different but then you have to transfer schematic to client and that is more complicated for limited benefit. Users of those commands are expected to have admin rights, thus also access to the server.

I was thinking, maybe a player would want to take a structure they built in another world and paste it one some multiplayer server or something like that. But it's fine with me if you don't want to support that.

Argmaster commented:
![image](https://github.com/user-attachments/assets/f0cb9542-4b5e-4f8e-8290-14d5ac22aa52)

[house.zip](https://github.com/user-attachments/files/19075515/house.zip)

Argmaster commented:
There is literally 0 error handling while loading tho, so there is still a bit to be done before making it ready for review

IntegratedQuantum commented:
Before you put any more work into it, I'd like to do some improvements to how other binary formats are handled, to improve error handling, among other things.

Argmaster commented:
> Before you put any more work into it, I'd like to do some improvements to how other binary formats are handled, to improve error handling, among other things.

How is that a bottle neck for this PR?

IntegratedQuantum commented:
> How is that a bottle neck for this PR?

Well you did say you want to implement more error handling, this is me telling that you should wait.

IntegratedQuantum commented:
Well it won't stall the PR, but since this is not going to finish review in the next weeks, you might as well wait with it, until I implement better options.

Argmaster commented:
>  this is not going to finish review in the next weeks

Good to know

Argmaster commented on src/blueprint.zig:
```diff
@@ -0,0 +1,338 @@
+	deflate,
+};
+
+/// Blueprint storage file format header structure.
+///
+/// To extend, create a packed struct `<something>HeaderExtension` and add it as last field to this struct.
+/// Bump `blueprintVersion` variable to indicate file format change.
+/// Add serialization logic for `<something>HeaderExtension` to `store` and `load` methods
+/// alongside version check to load that part of header only for files with version that supports it.
+/// If you are adding new data segment, add serialization and deserialization logic to
+/// `Blueprint.store` and `Blueprint.load` methods.
```
```suggestion

```

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+
+		const sizeX: usize = @intCast(@abs(endX - startX + 1));
+		const sizeY: usize = @intCast(@abs(endY - startY + 1));
+		const sizeZ: usize = @intCast(@abs(endZ - startZ + 1));
```
You don't need the intCast. `@abs` already produces an unsigned integer.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+				for(0..sizeZ) |offsetZ| {
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = main.server.world.?.getBlock(worldX, worldY, worldZ) orelse Block{.typ = 0, .data = 0};
```
If the chunk is not loaded, then we should return an error and message the user in the command feedback instead of silently failing.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = main.server.world.?.getBlock(worldX, worldY, worldZ) orelse Block{.typ = 0, .data = 0};
+					self.blocks.append(block);
```
Please use self.blocks.appendAssumeCapacity, and reserve the capacity up-front.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+
+		const sizeX: usize = self.sizeX;
+		const sizeY: usize = self.sizeY;
+		const sizeZ: usize = self.sizeZ;
```
I don't think it is particularly useful to have so many aliases of variables that only get used once.

Argmaster replied:
Side effect of redesigning the function many times - long time ago there was actual expression going with it.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+					const worldZ = startZ + @as(i32, @intCast(offsetZ));
+
+					const block = self.blocks.items[blockIndex];
+					mesh_storage.updateBlock(worldX, worldY, worldZ, block);
```
This will fail on headless servers.

Argmaster replied:
How do I check if server is headless? Or just how do I check if this is allowed?

IntegratedQuantum replied:
This is never allowed. You should never try to access client-side data structures from server side code.
In fact in this case this would even lead to updating the data twice, once through network interface, and once here.

Argmaster replied:
So should I just discard this line?

IntegratedQuantum replied:
yes

Argmaster replied:
```suggestion
```

By the way, just so you know, you can create directly committable code change suggestions in Github UI. You have to create code block with `suggestion` as language or use a button in UI, then It also copies current code automatically. You can select multiple lines too. I like this a lot when I'm doing reviews.

![image](https://github.com/user-attachments/assets/afe79971-e643-482a-bb8d-6947a22fca7d)

IntegratedQuantum replied:
Oh I didn't know that. I'll try that out.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+		}
+	}
+	pub fn store(self: @This(), externalAllocator: NeverFailingAllocator) []u8 {
+		const allocator = main.stackAllocator;
```
`allocator` is a name that's often used for what you call `externalAllocator` here.
In fact even in your own code you have used it.

Furthermore, as discussed before, I don't like that you alias the stackAllocator locally.

IntegratedQuantum reviewed (changes requested):
First partial review (sorry I don't have the patience to do this in one sitting)

Some things to think about:
- Should the positions even be stored on the server?
  In my opinion they should be client-only and shown in the GUI (by drawing a cuboid around the selected region, and marking the 2 selected corner blocks in separate colors) as well as edited graphically (instead of typing a command, I think right/left-clicking with a tool would be much nicer)
- Should we add 10 different commands, or would it be nicer to just have 1 command (e.g. `/bp`) which can do everything as a second argument (load/store/capture/place)

Argmaster commented:
> First partial review (sorry I don't have the patience to do this in one sitting)

Sure, actually didn't expect review for some more time.

> * Should the positions even be stored on the server?

No idea how to do it differently than right now, went with simple implementation. Since commands are server side I would think it should, but I have no strong opinion as long as sb gives me guideline how to implement this differently.

>   In my opinion they should be client-only and shown in the GUI (by drawing a cuboid around the selected region, and marking the 2 selected corner blocks in separate colors) 

Cool idea, but we are 700 lines already, visualization is a great thing for @OneAvargeCoder193 to pick up since he clearly enjoys improving graphical aspects of my implementations 😜 

> as well as edited graphically (instead of typing a command, I think right/left-clicking with a tool would be much nicer)

- using commands this is a well known pattern
- commands are reasonably easy to implement
- both command based and selection tool can coexist - in fact this is the intended design, mirroring what WorldEdit does.
- since this feature was intended as similar to WorldEdit, I'd like to keep commands.
- if headless servers have command line, having commands would allow placing stuff from there, hard to anticipate how cool that can be, but I would argue It is great.

> * Should we add 10 different commands, or would it be nicer to just have 1 command (e.g. `/bp`) which can do everything as a second argument (load/store/capture/place)

There are commands like `//stack`, `//smooth`, `//move` which operate on clipboard but have nothing to do with blueprints per-se. The fact that we use blueprint for clipboard is a implementation detail. I would prefer to keep general principles from WorldEdit as to not reinvent the wheel, ie. blueprints exist in user mind only for IO. **No one wants to learn yet another design of world editor.** 

That said, we could have `/we <subcommand>` (`/we blueprint save foo`) but this is a lot of extra writing, there is a reason why position setting is `/pos1` not `/position1` and this is also why WorldEdit uses `//` prefix for its commands instead of `/we` or sth.

Argmaster commented:
Also, box selection is not only selection mode in WorldEdit and I would like to implement at least some of the others, which means that we would probably need void blocks ie. air that do not replace anything.

IntegratedQuantum commented:
> both command based and selection tool can coexist - in fact this is the intended design, mirroring what WorldEdit does.

I don't think we should mindlessly mirror all of WorldEdit's design decision.
There a reason world edit has (clunky) commands instead of being a GUI-first editor: Its goal is to run in client-unmodded minecraft. Cubyz does not have that restriction.

I get that it is nice to have a familiar interface and familiar mechanics, but do people actually use these set pos commands for example? Doesn't everyone just run around with a wooden axe instead?

Argmaster commented:
> I get that it is nice to have a familiar interface and familiar mechanics, but do people actually use these set pos commands for example? Doesn't everyone just run around with a wooden axe instead?

I do, yes, rarely, but it still a thing, in certain situations much more convenient than using axe, eg. placing pos2 mid-air for big build.

IntegratedQuantum commented:
> I do, yes, rarely, but it still a thing, in certain situations much more convenient than using axe, eg. placing pos2 mid-air for big build.

Yeah, but again we are not bound by the same restrictions. We could e.g. implement it such that clicking in the air selects the current position.

Argmaster commented:
> Yeah, but again we are not bound by the same restrictions (I think worldedit only gets an event when a block is broken or interacted with). We could e.g. implement it such that clicking in the air selects the current position.

I think MCEdit had something like this, you clicked once to start selection and it dragged with you until you clicked second time and it was visible as semi-transparent white box. Also what you were aiming at was outlined. There were also different "aiming at" modes, one with capped distance and one with uncapped.

Regardless, I consider this outside of the scope of this PR and since I didn't work with our graphics code much it would really significantly defer this PR and possibly inflate it. I can't promise I will keep working on it if it takes a month to finish it.

On top of that I still think it is good to have commands for sake of editing from headless command line, if not for average player convenience. 

I do have to admit I am biased since I am used to how I use WorldEdit commands after using it thousands of times. But I am unlikely to be only one.

Argmaster commented:
MCEdit for reference:

![image](https://github.com/user-attachments/assets/f342fc50-643b-4ed3-9eea-01677ad6aa12)

![image](https://github.com/user-attachments/assets/d3e3022e-2ddd-4ab1-ba4d-a26573934fe9)

Argmaster commented on src/blueprint.zig:
```diff
@@ -0,0 +1,330 @@
+const vec = main.vec;
+const Vec3i = vec.Vec3i;
+
+const mesh_storage = main.renderer.mesh_storage;
```
```suggestion
```

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,329 @@
+	paletteBlockCount: u16 = 0,
+	blockArraySizeX: u16 = 0,
+	blockArraySizeY: u16 = 0,
+	blockArraySizeZ: u16 = 0,
```
I'm not sure about adding a separate struct with automagical conversions for the header.

So far (as you've probably seen while implementing the binary reader/writer everywhere), I generally prefer to just have the read/writeInts inline where the rest of the protocol is handled to avoid any magic.

But I'm open to discuss this.

Argmaster replied:
This has a really significant advantage: You can't miss adding a field serialization / deserialization code after adding a field and you can't make a mistake in order of reads. If given type is already covered, then it will read it automagically, and if it is not it is runtime error (unfortunately not compile time, as zig requires else in this switch). 

This also creates quite easy to implement path for adding new segments, as they can be done as nested struct with dedicated case in switch. 

That said, its not a huge gain and having extensions to file header as nested struct is not actually that good, code will be deeply nested and accessing it will be weird. Having consistent code is nearly equally nice. 

Except the fact that for most of the cases of serialization-deserialization there are just two identical series of reads and writes that you have to manually keep in sync with no checks in place to confirm you didn't mess anything up.

What could make a difference is if we were to implement struct serialization like this at BinaryReader/BinaryWriter level (or at least as universal class) - so you an just pass a struct as comptime param and it will automatically read and validate it. This is one of core things introspection is used for - so you have a single source of truth for reading and writing.

Version field would probably have to be excluded from this and what you could do is have different structs for different versions. But this would save quite some code assuming you would make stuff into structs in the first place.

Having introspection based serialization just for `FileHeader` may not be worth it, as doing it manually may even be less code and definitely simpler. But doing that everywhere would significantly improve robustness of code (IMO).

IntegratedQuantum replied:
> You can't miss adding a field serialization / deserialization code after adding a field

Such a mistake can be easily catched when testing the new field.

> you can't make a mistake in order of reads/writes

Oh yes you can, and in a very subtle way: If you forget to make the struct packed or extern, than the field order is not guaranteed.

> But this would save quite some code assuming you would make stuff into structs in the first place.

Well even if you do make stuff into structs, it only matters if you store the structs. If you throw it away after storing/loading (as done here), then using local variables would be simpler than defining a struct, putting the values into the struct and getting them out of it again at the end.

> But doing that everywhere would significantly improve robustness of code (IMO).

Yes, it is slightly more robust, but in my opinion it encourages wastefulness (e.g. if you have a struct with some extra fields, then you are encouraged to just put it all in the file, instead of only extracting the data you actually need) and it also makes it harder to employ simple compression methods, such as varints, since you'd have to take the step back to manual serialization first.

Argmaster replied:
> Such a mistake can be easily catched when testing the new field.

You don't need to test code you didn't write.

> Oh yes you can, and in a very subtle way: If you forget to make the struct packed or extern, then the field order is not guaranteed.

Yes, but you create struct once, while you are likely to change its fields many times. Every time you do that you need to change serialization and deserialization code too.
 
> Well even if you do make stuff into structs, it only matters if you store the structs. If you throw it away after storing/loading (as done here), then using local variables would be simpler than defining a struct, putting the values into the struct and getting them out of it again at the end.

I don't necessarily agree. This is only ~33% code more if you use struct only for serialization/deserialization and have to move everything in/out, while in return you get compile time checking if you assigned all the fields correctly and single source of truth for order of storing fields.

> Yes, it is slightly more robust, but in my opinion it encourages wastefulness (e.g. if you have a struct with some extra fields, then you are encouraged to just put it all in the file, instead of only extracting the data you actually need)

For me current model encourages copying code, which generates higher maintenance cost, which has to be paid by humans, while waste of few bits is paid by machines, which really don't mind few more bits considering throughputs of networks and drives read Gb/s. If measured that performance penalty exists, it is reasonable to replace automated code with manual, but without that wastefulness of suggested solution is not proven hence not a convincing argument for me.

After all you are the one to review the code, so now instead of focusing on whether fields are in correct order you can focus on whether the design is not wasteful.

> and it also makes it harder to employ simple compression methods, such as varints, since you'd have to take the step back to manual serialization first.

Where exactly do you currently apply this kind of optimizations? In code I have refactored I don't recall any of these. On top of that, wouldn't it be possible to implement this in automated manner in struct serialization code (in a generic way!).

This is a solution for simple cases, and yes, there are always exceptions.

Imo this could be seamlessly applied in many places in code:

```zig
			const pos = chunk.ChunkPosition{
				.wx = try reader.readInt(i32),
				.wy = try reader.readInt(i32),
				.wz = try reader.readInt(i32),
				.voxelSize = try reader.readInt(u31),
			};
```
```zig
			const x = try reader.readInt(i32);
			const y = try reader.readInt(i32);
			const z = try reader.readInt(i32);
			const newBlock = Block.fromInt(try reader.readInt(u32));
```

```zig
			const pos = [_]f64{
				try reader.readFloat(f64),
				try reader.readFloat(f64),
				try reader.readFloat(f64),
				@floatCast(try reader.readFloat(f32)),
				@floatCast(try reader.readFloat(f32)),
				@floatCast(try reader.readFloat(f32)),
			};
			const vel = [_]f64{
				try reader.readFloat(f64),
				try reader.readFloat(f64),
				try reader.readFloat(f64),
				0,
				0,
				0,
			};
```

```zig
			var writer = utils.BinaryWriter.initCapacity(main.stackAllocator, networkEndian, 62);
			defer writer.deinit();
			writer.writeInt(u64, @bitCast(playerPos[0]));
			writer.writeInt(u64, @bitCast(playerPos[1]));
			writer.writeInt(u64, @bitCast(playerPos[2]));
			writer.writeInt(u64, @bitCast(playerVel[0]));
			writer.writeInt(u64, @bitCast(playerVel[1]));
			writer.writeInt(u64, @bitCast(playerVel[2]));
			writer.writeInt(u32, @bitCast(game.camera.rotation[0]));
			writer.writeInt(u32, @bitCast(game.camera.rotation[1]));
			writer.writeInt(u32, @bitCast(game.camera.rotation[2]));
			writer.writeInt(u16, time);
			conn.sendUnimportant(id, writer.data.items);
```

IntegratedQuantum replied:
> Yes, but you create struct once, while you are likely to change its fields many times. Every time you do that you need to change serialization and deserialization code too.

I think a little more friction when changing code could even be helpful here. It should make you stop and think "Wait, I'm changing something that gets stored. I should test if this still works for older save files."

> throughputs of networks and drives read Gb/s

Drives yes, networks definitely not. I just measured that I have an upload speed of 20 Mbit/s, and if I host a server and 10 people join, each one only gets effectively 2 Mbit/s, just barely enough for 500 item drops lying around, even less normal entities. Network is one of the most expensive resources, and it shouldn't be wasted for the sake of cleaner code.

> Where exactly do you currently apply this kind of optimizations?

I do use it only once in the network system so far to encode packet lengths. But I would like to use it more often. E.g. in this very PR the block id length would be a perfect target for varints. Most ids are less than 127 bytes and we could save a significant portion of the palette memory usage (which will likely be the dominant factor for small structures).

> Having functions for serialization of structs, vectors and arrays in universal manor would remove a lot of code.

Oh yes, serializing vectors would be great. But to be honest, for commonly used structs such as `ChunkPosition` we could also just add a member function instead of doing all the automated magic.

Argmaster replied:
> I think a little more friction when changing code could even be helpful here. It should make you stop and think "Wait, I'm changing something that gets stored. I should test if this still works for older save files."

This is not any sort of direct implication, unless when you modify other parts of the code that are hard to work with you also think "Oh hey this might require backwards compatibility". This though comes from you knowing that it is requirement, not from the fact that the code is hard to modify, since it is actually not, this is very simple code to modify, much harder to keep in sync, but the latter part is something you do unintentionally.
  
> Drives yes, networks definitely not. I just measured that I have an upload speed of 20 Mbit/s, and if I host a server and 10 people join, each one only gets effectively 2 Mbit/s, just barely enough for 500 item drops lying around, even less normal entities. Network is one of the most expensive resources, and it shouldn't be wasted for the sake of cleaner code.

Ok, possible.
 
> I do use it only once in the network system so far to encode packet lengths. But I would like to use it more often. 

Can you point me to that code? I can't seem to find it.

> E.g. in this very PR the block id length would be a perfect target for varints. Most ids are less than 127 bytes and we could save a significant portion of the palette memory usage (which will likely be the dominant factor for small structures).

I seriously doubt you can gain anything from that since palette and block array are compressed with deflate which already does bit reduction and sequence aliasing. If having that particular value represented by shorter sequence of bytes had significant impact on file size, deflate would most likely notice that. This might be a reasonable optimization in places where data is to short to be effectively compressed or compression may take too long for it to be useful, which leaves us probably only with networking.

> > Having functions for serialization of structs, vectors and arrays in universal manor would remove a lot of code.
> 
> Oh yes, serializing vectors would be great. But to be honest, for commonly used structs such as `ChunkPosition` we could also just add a member function instead of doing all the automated magic.

Honestly, I don't agree with calling it magic. It is introspection, its nowhere close to a real magic, like sick stuff you can do with Python or Rust macros, in the first one you can literally change everything, every object, as you go, in latter you can literally write code in different language within Rust file. This is magic. What we have here is simple introspection, and it is here  for the purpose of not writing serialization functions for individual classes if they are trivially serializable (and many other I guess).

IntegratedQuantum replied:
> Can you point me to that code? I can't seem to find it.

https://github.com/PixelGuys/Cubyz/blob/e7ecd161ce00851240ad5bb9bcb1aa0bd69dc1d5/src/network.zig#L1421-L1426
https://github.com/PixelGuys/Cubyz/blob/e7ecd161ce00851240ad5bb9bcb1aa0bd69dc1d5/src/network.zig#L1641-L1658

> I seriously doubt you can gain anything from that since palette and block array are compressed with deflate which already does bit reduction and sequence aliasing. 

Well, for chunk data it reduced storage by 30% just by storing 1 byte instead of 4 in the deflate input.
Now of course here it would be less relevant. But if implemented in the binary reader it adds no complexity to the code, but gains a small reduction in file size, so why not use it?

> Honestly, I don't agree with calling it magic. It is introspection

I call it magic because automatic serialization has to make some difficult decisions (how to deal with pointers? How to deal with strings in particular? Should override functions exist?), and these decisions can have unexpected side effects which are not at all obvious to the outside (e.g. consider serializing a MultiArrayList, which internally uses a byte array).

I personally think that it's easy to spot mistakes in the simple and explicit serialization code, whereas the magic black box can have unforeseen consequences which may haunt us only years later. Let's say I change the fields of `ChunkPosition`, and oops some old world files break. And the relevant serialization code doesn't even show up in the review either, so if we find out about it 5 PRs later then it will be a hell of a debugging session.
With explicit code the compiler will immediately complain about it, and any changes to that code path will also show up in the review, so there is no way such a broken change could get merged.

Argmaster replied:
Ok, I guess let's agree to disagree and do it your way. I assume this creates requirement to rewrite it to manual serialization.

Argmaster replied:
Awesome.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+	blocks: main.List(Block),
+	sizeX: usize,
+	sizeY: usize,
+	sizeZ: usize,
```
I think `utils.Array3D` would be better here.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+
+		const sizeX: usize = @abs(endX - startX + 1);
+		const sizeY: usize = @abs(endY - startY + 1);
+		const sizeZ: usize = @abs(endZ - startZ + 1);
```
Should use `@intCast` instead of `@abs`, since we know that end is always larger than start.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+		std.debug.assert(decompressedDataSizeBytes != 0);
+
+		var decompressedWriter = BinaryWriter.initCapacity(allocator, .big, decompressedDataSizeBytes);
+		defer decompressedWriter.deinit();
```
`decompressed` implies that it was compressed before, which is not true.
Maybe `uncompressedWriter` or `rawWriter` would be better?

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+		const decompressedDataSizeBytes = blockPaletteSizeBytes + blockArraySizeBytes;
+		std.debug.assert(decompressedDataSizeBytes != 0);
+
+		var decompressedWriter = BinaryWriter.initCapacity(allocator, .big, decompressedDataSizeBytes);
```
For local allocations, that are freed at the end of the scope, always use the stackAllocator.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+		self.storeBlockArray(&decompressedWriter, gameIdToBlueprintId);
+
+		const compression: BlueprintCompression = .deflate;
+		const compressedData: []u8 = self.compressOutputBuffer(main.stackAllocator, decompressedWriter.data.items, compression);
```
In my opinion, the compression function should be the one deciding what compression to use, and then return it from the function.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+		const compression: BlueprintCompression = .deflate;
+		const compressedData: []u8 = self.compressOutputBuffer(main.stackAllocator, decompressedWriter.data.items, compression);
+		defer main.stackAllocator.free(compressedData);
+		std.debug.assert(compressedData.len != 0);
```
0 bytes of data can be a valid result of a compression algorithm. The decompression function could then e.g. just fill the entire output buffer with 0 (air).

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+		defer main.stackAllocator.free(blockPalette);
+		std.debug.assert(gameIdToBlueprintId.count() == blockPalette.len);
+
+		const blockPaletteSizeBytes = self.getBlockPaletteSize(blockPalette);
```
Instead of adding a function for this, I think this should just use the writer length after the data has been written to it.

Argmaster replied:
`blockPaletteSizeBytes` is used to calculate how big the allocation for BinaryWriter should be, so based on your suggestion we prefer reallocation of writer buffer instead of a function that precalculates that value?

IntegratedQuantum replied:
Reallocations for the stackAllocator cost practically nothing if it's the top-most item on the stack (which it is here)

Argmaster replied:
Ok

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+			writer.writeInt(BlockIdSizeType, @intCast(blockName.len));
+			writer.writeSlice(blockName);
+		}
+	}
```
I personally would reorder the functions, such that they are grouped in load/store pairs.
Furthermore for every store helper function I'd add a symmetric load function.

IntegratedQuantum replied:
I fail to understand how the current version is more readable than what I proposed.
I'd argue, that when this code is read, it's either because I'm looking for a bug (and there it is helpful when serialization/deserialization code are next to each other), or it's because someone wants to add a new field or something (→ modifying the code).

Argmaster replied:
- With current layout you can start at the top and scroll down to dive deeper into the details, understanding the flow from most abstract to least abstract level. [This is a natural way to read text.](https://github.com/Geeksltd/Programming.Tips/blob/master/docs/methods/stepdown-rule.md)
- If you want to compare loading and storing side by side, just open second editor to the side. Since functions are in same order, now you have double gain from current grouping and instead of having interleaved functions, you have two clean series of functions.
- To find a bug you need to first understand the flow of the code. This means reading the code first.
- You will read this code to decide if you need to modify it, add field or whatever.
- You will need to read this code second time after you modify it to make sure what you did makes sense in the flow of the code.

Primary navigation principle here is to keep things that are directly dependent on each other together, rather than keep things that are indirectly dependent on each other together. Implementation details of `load` are directly dependent on each other. `store` and `load` are indirectly dependent - they must match because they are two sides of same coin.

Argmaster replied:
However, I still think that having to manually keep serialization and deserialization in sync is a sick idea and order of function doesn't really make this less tedious.

IntegratedQuantum replied:
> With current layout you can start at the top and scroll down to dive deeper into the details, understanding the flow from most abstract to least abstract level. [This is a natural way to read text.](https://github.com/Geeksltd/Programming.Tips/blob/master/docs/methods/stepdown-rule.md)

This concept is not hindered by grouping the functions together, you could still keep the high-level load/store at the top with their helpers underneath.

> If you want to compare loading and storing side by side, just open second editor to the side.

Opening a new tab in the editor obstructs the flow of reading.
And I do in fact sometimes use multiple tabs when reviewing code, but it always gets messy and I accidentally scroll in the wrong tab and things like that.

-------------------------

It seems that we have just fundamentally different approaches to reading code.
I mainly use go-to-definition when navigating the code, which makes your layout almost useless.
When direct dependencies are handled by the IDE, it's more important to have indirect dependencies (which cannot be visualized by the IDE) visualized in the code.

Argmaster replied:
> It seems that we have just fundamentally different approaches to reading code.
> I mainly use go-to-definition when navigating the code, which makes your layout almost useless.
> When direct dependencies are handled by the IDE, it's more important to have indirect dependencies (which cannot be visualized by the IDE) visualized in the code.

Well, If I were to guess, this may come from the fact you are accustomed to this codebase, have better understanding of dependencies and possibly better memory to keep track of stuff, than I do. For me going through this code with no consistent structure is hard, as there is no simple principle where look for stuff (and how to go back after that) while I have to consistently go back and forth to refresh what is happening and why. 

On top of that I consider things like go-to-definition (if they even work) to be a secondary way to move around, when I can't find something I am looking for nearby. Go-to often works well one way - you can go from use to definition - but if you try to go from definition back to use it is likely you will have a list of many uses to chose from which means there is no easy way to go back (yeah, not entirely true, there is a shortcut for that I never remember, for sure there is `alt` + `<-` to move cursor to previous position), scrolling is just easier and faster for me most of the time, maybe because I am used to having one finger on scroll / page-up page-down keys.

As for scrolling wrong tab, yeah, well, it happens, but you can just scroll the same amount back and you are good, with go to its a bit more tricky. If you jump 500 lines to a definition then going back is just a bit more inconvenient.

IntegratedQuantum replied:
Well in my opinion it's definitely worth it to learn the shortcut to go back. I also only learned it recently, and honestly it's so much better than just scrolling up and down all the time.

> go-to-definition (if they even work)

Yeah ZLS tends to have trouble with following `@import("root")`, it confuses which root to take when there are multiple projects in the build.zig. I had to disable the formatters in the build.zig to make ZLS work for me locally.

---------

Either way this discussion doesn't seem to lead anywhere. So I'd propose the following:
How about you just try it, moving the respective store and load next to each other. Then you can see how it feels, and if it's really that much harder to navigate for you, then we can do it your way.

Argmaster replied:
It's not the first time I am doing code, not the first time I do serialization. We will do it your way cause its your project and that's the end of discussion indeed.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,281 @@
+
+		return gameIdToBlueprintId;
+	}
+	fn packBlockPalette(_: @This(), allocator: NeverFailingAllocator, map: GameIdToBlueprintIdMapType) [][]const u8 {
```
if you don't use self, then you shouldn't take it as a parameter in my opinion.

Argmaster replied:
Probably, I'm used to Python where it is always present unless marked `@staticmethod` or `@classmethod`.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+
+	var splitIterator = std.mem.splitSequence(u8, args, " ");
+	while(splitIterator.next()) |a| {
+		argsList.append(main.stackAllocator.dupe(u8, a));
```
duping is not needed, since `args` (which `a` points into) lives for the entire duration of the function.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+pub const description = "Operate on blueprints.";
+pub const usage = "/blueprint <save FILENAME|delete FILENAME|load FILENAME|list>";
```
Wouldn't it be more readable as separate lines? (Also quick reminder to avoid capslock)
```
/blueprint save <file name>
/blueprint delete <file name>
/blueprint load <file name>
/blueprint list
```
This is how other commands deal with subcommands in the help menu.

Argmaster replied:
Caps lock here is for different reason than in variable names, it was introduced as part of unix command line argument naming convention where positional arguments are noted in caps lock.

```
$ ls --help
Usage: ls [OPTION]... [FILE]...
List information about the FILEs (the current directory by default).
Sort entries alphabetically if none of -cftuvSUX nor --sort is specified.
``` 

I'll change it as requested.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+
+	var cwd = std.fs.cwd();
+
+	_ = cwd.makeDir("blueprints") catch null;
```
Should use `try` as well, like the command below it.

(Also instead `_ = ... catch null` you could do `... catch {}` to return a void type.)

Argmaster replied:
Doesn't it fail when path already exist?

IntegratedQuantum replied:
Oh right it does. Either way you should use main.files
(Or at least openPath (which creates the directory automatically) instead of openDir)

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+		try blueprintsDir.writeFile(.{
+			.sub_path = fileName,
+			.data = storedBlueprint,
+			.flags = .{.lock = .exclusive},
```
Please don't silently introduce this into your PR. If this is fixing a relevant issue, then it should be discussed in an issue first.

IntegratedQuantum replied:
Also quick reminder that main.files exist, so that we don't need to do all these incantations every time we want to read or write a file. e.g. `main.files.openDir` makes `makeDir` obsolete. Calling `.write()` on the result then allows you to write a file directly.

Argmaster replied:
Well, It is preventing an issue, not fixing it, I don't like judging synchronization issues as they kick back when you least expect. If the cost of synchronization is low (which it is in this case) It should be done imo. This is relevant only if at any point of existence of the game commands can or will be able to be executed independently for different players. That said, maybe the best approach would be to have sub folders for each player for their own set of blueprints (created when first blueprint is saved), after all even if blueprint files is not corrupted, having someone else overwrite your stuff is not a fun option.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+		defer blueprintsDir.close();
+
+		std.log.info("Saving clipboard to blueprint file: {s}", .{fileName});
+		source.sendMessage("#00ff00Saving clipboard to blueprint file: {s}", .{fileName});
```
In my opinion the result should be sent and logged after the saving was successful, not before.

Argmaster replied:
During debug it was more helpful for me to know that we reached point before call rather, but I guess for future having it after will be better.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```
The lock should be around only the necessary bits of data (→the `source.commandData.clipboard != null` check and writing the new blueprint to the clipboard).
This makes it harder to run into deadlocks (like what if `source.sendMessage` also needs to lock the mutex in the future?) and it's faster too.

IntegratedQuantum commented on src/server/command/worldedit/copy.zig:
```diff
@@ -0,0 +1,44 @@
+		source.sendMessage("#ff0000Too many arguments for command /copy. Expected no arguments.", .{});
+		return;
+	}
+	source.mutex.lock();
```
Same here, the lock should only be taken in the success branch. And the old clipboard should only be discarded in the success branch

Argmaster replied:
How do you like the Result structure pattern? Yoinked it from Rust.

IntegratedQuantum replied:
> How do you like the Result union/enum pattern (capture)? Yoinked it from Rust.

I  don't particularly like it, but I don't really see a better way to implement it either.

IntegratedQuantum commented on src/server/command/worldedit/deselect.zig:
```diff
@@ -0,0 +1,21 @@
+	source.mutex.lock();
+	defer source.mutex.unlock();
+
+	source.commandData.selectionPosition2 = null;
```
Shouldn't this also deselect pos1? At least the command description says so.

Argmaster replied:
Indeed

IntegratedQuantum commented on src/server/command/worldedit/deselect.zig:
```diff
@@ -0,0 +1,21 @@
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```
Also same problem with the mutex.

IntegratedQuantum commented on src/server/command/worldedit/pos1.zig:
```diff
@@ -0,0 +1,27 @@
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```
same problem with the mutex.

IntegratedQuantum commented on src/server/command/worldedit/pos2.zig:
```diff
@@ -0,0 +1,27 @@
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```
same problem with the mutex.

IntegratedQuantum commented on src/server/command/worldedit/showpos1.zig:
```diff
@@ -0,0 +1,22 @@
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```
same problem with the mutex.

IntegratedQuantum commented on src/server/command/worldedit/showpos2.zig:
```diff
@@ -0,0 +1,22 @@
+		return;
+	}
+	source.mutex.lock();
+	defer source.mutex.unlock();
```
same problem with the mutex.

IntegratedQuantum commented on src/server/server.zig:
```diff
@@ -18,6 +18,18 @@ pub const storage = @import("storage.zig");
 
 const command = @import("command/_command.zig");
 
+pub const UserCommandData = struct {
```
Wouldn't it be better to call it `WorldEditData`?

Argmaster replied:
I wanted to prepare a place for future command state if it ever exists, it doesn't really make sense to have every bit of command state separately IMO, a lot of potential additional init and deinit calls would be required.

IntegratedQuantum replied:
Well, in my opinion we shouldn't sacrifice readability for an uncertain future.
And if we do discover that many commands need user state, then we should probably solve this in a more generic way, such that mods (if they'll ever be a thing) can also use it.

Argmaster replied:
As you prefer, although I don't think it is significantly harming readability - it is not a lie, just the name is a umbrela term.

> And if we do discover that many commands need user state, then we should probably solve this in a more generic way, such that mods (if they'll ever be a thing) can also use it.

Definitely outside of the scope of your PR.

Argmaster commented on src/renderer.zig:
```diff
@@ -941,5 +941,13 @@ pub const MeshSelection = struct { // MARK: MeshSelection
+			if(game.Player.selectionPosition2) |pos2| {
+				c.glEnable(c.GL_POLYGON_OFFSET_LINE);
+				defer c.glDisable(c.GL_POLYGON_OFFSET_LINE);
+				c.glPolygonOffset(-2, 0);
```
@OneAvargeCoder193 could you help?

OneAvargeCoder193 replied:
this was copied from the block selection code, why would the block selection need it but not the worldedit selection?

Argmaster replied:
There is also a bug with selection:

Displayed selection:

![image](https://github.com/user-attachments/assets/d6b64824-fca6-461d-9724-0bc8d553d931)

Actual selection:

![image](https://github.com/user-attachments/assets/bf2e1b28-33ab-4686-893f-f968853d690c)

IntegratedQuantum replied:
Oh I thought you were just drawing a line, nevermind on the offset then.

> There is also a bug with selection:

Yeah, the code doesn't handle the case that pos2 is larger than pos1 correctly.

Argmaster replied:
Kind of fixed it.

Argmaster commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,193 @@
+	};
+}
+
+fn blueprintSave(args: List([]const u8), source: *User) !void {
```
Well, many times in this case is up to 3 per function, after calling `self.blocks.width` *so many times* due to use of `Array3D` I don't feel like it is much 😝 
I'll change it tho

Argmaster replied:
It's probably because I am still not used to arrays (slices) knowing their size.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -1,17 +1,24 @@
 const std = @import("std");
 
 const main = @import("root");
+//const main = @import("../../../main.zig");
```
Unintended change

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -81,24 +83,17 @@ fn blueprintSave(args: List([]const u8), source: *User) !void {
-		_ = cwd.makeDir("blueprints") catch null;
-
-		var blueprintsDir = try cwd.openDir("blueprints", .{});
+		std.fs.cwd().makeDir("blueprints") catch {};
```
No longer necessary, since main.files.openDir already takes care of it.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -112,41 +107,35 @@ fn ensureBlueprintExtension(allocator: NeverFailingAllocator, fileName: []const
-	_ = cwd.makeDir("blueprints") catch null;
-
-	var blueprintsDir = try cwd.openDir("blueprints", .{});
+	std.fs.cwd().makeDir("blueprints") catch {};
```
same here

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -112,41 +107,35 @@ fn ensureBlueprintExtension(allocator: NeverFailingAllocator, fileName: []const
-	var blueprintsDir = try cwd.openDir("blueprints", .{.iterate = true});
+fn blueprintList(source: *User) !void {
+	std.fs.cwd().makeDir("blueprints") catch {};
+	var blueprintsDir = try openDir("blueprints");
```
Use makeOpenPath instead of makeDir+openDir

IntegratedQuantum replied:
Also in order to iterate it, you need to set the iterate option, which is not possible with main.files.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -158,35 +147,32 @@ fn blueprintList(_: List([]const u8), source: *User) !void {
-	_ = cwd.makeDir("blueprints") catch null;
-
-	var blueprintsDir = try cwd.openDir("blueprints", .{});
+	std.fs.cwd().makeDir("blueprints") catch {};
```
makeDir is redundant, like above

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -158,35 +147,32 @@ fn blueprintList(_: List([]const u8), source: *User) !void {
 	if(source.commandData.clipboard != null) {
 		source.commandData.clipboard.?.deinit(main.globalAllocator);
 	}
 	source.commandData.clipboard = try Blueprint.load(main.globalAllocator, storedBlueprint);
```
Could you also move the load outside the mutex?

IntegratedQuantum commented on src/server/command/worldedit/paste.zig:
```diff
@@ -19,9 +19,10 @@ pub fn execute(args: []const u8, source: *User) void {
 	source.mutex.lock();
-	defer source.mutex.unlock();
+	const nullableClipboard = source.commandData.clipboard;
+	source.mutex.unlock();
```
Another thread may deinit the clipboard while we still hold a reference to it. The lock is too tight here.

IntegratedQuantum reviewed (changes requested):
Also I wonder: Why do we even need locks in the first place?
The worldedit state is only edited by commands, and commands (of one player) must be executed sequentially anyways, to ensure they are executed in the same order that player sent them.

Argmaster commented:
> Also I wonder: Why do we even need locks in the first place? The worldedit state is only edited by commands, and commands (of one player) must be executed sequentially anyways, to ensure they are executed in the same order that player sent them.

Maybe we don't, I don't know that, I don't know all the implementation details, I saw you are locking accesses to player data so I followed.

IntegratedQuantum commented on src/zon.zig:
```diff
@@ -250,6 +250,10 @@ pub const ZonElement = union(enum) { // MARK: Zon
 
+	pub fn append(self: *const ZonElement, value: anytype) void {
+		self.array.append(createElementFromRandomType(value, self.array.allocator.allocator));
+	}
```
Please rebase the PR, and remove this function, since it was just added in #1190

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+	var splitIterator = std.mem.splitScalar(u8, args, ' ');
+	while(splitIterator.next()) |a| {
+		argsList.append(a);
+	}
```
Actually that this is not really the way the split iterator is intended to be used.
Ideally you'd avoid ever storing a list of arguments, and the function you call would just (attempt to) consumes the next item of the iterator.
Well, but now you've implemented it like this, and it's not really a huge deal so I won't force you to change this all again, it's probably not a big deal here anyways.
I just wanted to let you know what the intended usage is. Zig and its standard library are trying to guide you towards the simpler and faster choices, but it's not always quite clear what to do.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+			source.sendMessage("#ff0000Unrecognized subcommand for /blueprint: '{s}'", .{argsList.items[0]});
+		},
+		.empty => {
+			source.sendMessage("#ff0000Missing subcommand for **/blueprint**, usage: {s} ", .{usage});
```
Why does this one make the blueprint bold, but the previous ones don't?

Argmaster replied:
At first I though it is a good idea to have them bold, but it would be inconsistent with other commands so I abandoned it and apparently didn't clean up properly.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+		return;
+	}
+	const subcommand = BlueprintSubCommand.fromString(argsList.items[0]);
+	_ = switch(subcommand) {
```
Do we need the `_ = `? Usually if you need it, it's the sign of a design flaw.

Argmaster replied:
It allows me to do generic catch on this whole switch, otherwise I would have to have separate catch in each case.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+			source.sendMessage("#ff0000Missing subcommand for **/blueprint**, usage: {s} ", .{usage});
+		},
+	} catch |err| {
+		std.log.info("Error: {s}", .{@errorName(err)});
```
This is part of the log and should give a more specialized message.
Otherwise it will be impossible to tell where this came from.
Furthermore, should this maybe be a warning instead of an info?

Argmaster replied:
Let's just move error handling directly to source of errors.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+
+fn blueprintSave(args: []const []const u8, source: *User) !void {
+	if(args.len < 2) {
+		source.sendMessage("#ff0000**/blueprint save** requires FILENAME argument.", .{});
```
```suggestion
		source.sendMessage("#ff0000**/blueprint save** requires file-name argument.", .{});
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+		return;
+	}
+	if(args.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint save**. Expected 1 argument, FILENAME.", .{});
```
```suggestion
		source.sendMessage("#ff0000Too many arguments for **/blueprint save**. Expected 1 argument, file-name.", .{});
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+
+fn blueprintDelete(args: []const []const u8, source: *User) !void {
+	if(args.len < 2) {
+		source.sendMessage("#ff0000**/blueprint delete** requires FILENAME argument.", .{});
```
```suggestion
		source.sendMessage("#ff0000**/blueprint delete** requires file-name argument.", .{});
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+		return;
+	}
+	if(args.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint delete**. Expected 1 argument, FILENAME.", .{});
```
```suggestion
		source.sendMessage("#ff0000Too many arguments for **/blueprint delete**. Expected 1 argument, file-name.", .{});
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+		return;
+	}
+	if(args.len >= 3) {
+		source.sendMessage("#ff0000Too many arguments for **/blueprint load**. Expected 1 argument, FILENAME.", .{});
```
```suggestion
		source.sendMessage("#ff0000Too many arguments for **/blueprint load**. Expected 1 argument, file-name.", .{});
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+	defer main.stackAllocator.free(fileName);
+
+	var blueprintsDir = try openDir("blueprints");
+	defer blueprintsDir.close();
```
I keep seeing the same thing three times here. I wonder if it would be possible to turn it into a helper function?
Probably not, but maybe you have an idea how it could be done?

Argmaster replied:
I am not sure which parts you mean. For `fileName` and `blueprintDir` alone I don't see benefit in extracting them to separate function, it would be almost the same amount of code, you still need to create both, return both and clean up both. As for parameter list checking, this is coincidental, those functions can in the future accept different flags. Ideally this would be abstracted by command parsing library, maybe a home grown one so it makes good use of source.sendMessage.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+		if(entry.kind != .file) break;
+		if(!std.ascii.endsWithIgnoreCase(entry.name, ".blp")) break;
+
+		source.sendMessage("#ffffff{}#00ff00 {s}", .{index, entry.name});
```
Can the index be used anywhere? If not, then I would not print it.

Argmaster replied:
Ideally you would be able to click on message in chat to load particular model. Index itself is as good as a `-` or anything else you could use to signalize that this is a list. Is there a reason to prefer unordered list over ordered one?

Argmaster replied:
Ok, replace them with dashes.

IntegratedQuantum replied:
In my opinion they also shouldn't be green.

Argmaster replied:
Ok

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,170 @@
+	defer main.stackAllocator.free(storedBlueprint);
+
+	if(source.commandData.clipboard != null) {
+		source.commandData.clipboard.?.deinit(main.globalAllocator);
```
Why not use a capture?
```zig
	if(source.commandData.clipboard) |oldClipboard| {
		oldClipboard.deinit(main.globalAllocator);
```

Argmaster replied:
`deinit` takes mutable pointer, I guess it doesn't have to.

IntegratedQuantum commented on src/renderer.zig:
```diff
@@ -941,5 +941,24 @@ pub const MeshSelection = struct { // MARK: MeshSelection
+				const bottomLeft: Vec3i = .{
+					@min(pos1[0], pos2[0]),
+					@min(pos1[1], pos2[1]),
+					@min(pos1[2], pos2[2]),
```
`@min` can operate on vectors

IntegratedQuantum commented on src/renderer.zig:
```diff
@@ -941,5 +941,24 @@ pub const MeshSelection = struct { // MARK: MeshSelection
+					@max(pos1[0], pos2[0]),
+					@max(pos1[1], pos2[1]),
+					@max(pos1[2], pos2[2]),
+				};
```
same here

IntegratedQuantum commented on src/network.zig:
```diff
@@ -999,7 +1006,27 @@ pub const Protocols = struct {
+							const pos = Vec3i{
+								@bitCast(try reader.readInt(i32)),
+								@bitCast(try reader.readInt(i32)),
+								@bitCast(try reader.readInt(i32)),
```
Why bitCast i32 to i32?

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,254 @@
+	pub fn clone(self: *Blueprint, allocator: NeverFailingAllocator) Blueprint {
+		var new = Blueprint.init(allocator);
+		new.blocks = self.blocks.clone(allocator);
+		return new;
```
Why not just `return .{.blocks = self.blocks.clone(allocator)}`. Then you'd also get an error here if add a new field in the future.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,254 @@
+				for(0..sizeZ) |z| {
+					const worldZ = startZ +% @as(i32, @intCast(z));
+
+					const maybeBlock = main.server.world.?.getBlock(worldX, worldY, worldZ);
```
Ideally you would get the chunk, and then add all blocks from it in one go (which would require 6 loops).
However I can understand that you may not want to go through the trouble, and at the moment that is far from becoming a bottleneck (especially for pasting)

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,254 @@
+	};
+
+	pub fn capture(allocator: NeverFailingAllocator, pos1: Vec3i, pos2: Vec3i) CaptureResult {
+		var self = Blueprint{.blocks = undefined};
```
self isn't used before `self.blocks` is assigned, and it is never modified afterwards, so why not declare it below as a const?

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,254 @@
+		}
+	}
+	pub fn load(allocator: NeverFailingAllocator, inputBuffer: []u8) !Blueprint {
+		var self = Blueprint{.blocks = undefined};
```
same here

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,254 @@
+		const paletteBlockCount = try compressedReader.readInt(u16);
+		const width: u32 = @intCast(try compressedReader.readInt(u16));
+		const depth: u32 = @intCast(try compressedReader.readInt(u16));
+		const height: u32 = @intCast(try compressedReader.readInt(u16));
```
casting is not needed, integers automatically coerce to bigger types.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,254 @@
+		const blockPaletteSizeBytes = storeBlockPalette(main.stackAllocator, gameIdToBlueprintId, &uncompressedWriter);
+
+		for(self.blocks.mem) |block| {
+			const blueprintBlock: BlockStorageType = (Block{.typ = gameIdToBlueprintId.get(block.typ).?, .data = block.data}).toInt();
```
How about
```suggestion
			const blueprintBlock: BlockStorageType = Block.toInt(.{.typ = gameIdToBlueprintId.get(block.typ).?, .data = block.data});
```

IntegratedQuantum commented:
Oh also, the suggestions to change the commands to `/pos1 show` and `/pos1 clear` are not a requirement if you don't want to.

Argmaster commented on src/server/command/worldedit/deselect.zig:
```diff
@@ -0,0 +1,20 @@
+const User = main.server.User;
+
+pub const description = "Clears pos1 and pos2 of selection.";
+pub const usage = "/deselect";
```
No, since for setting it is `/pos1` not `/pos1 set`, and the latter is awkward since `set` might suggest there is another parameter needed. Also there is already precedent set by WorldEdit, if anyone is going consider any approach more intuitive than other I am farily certain it will be the WorldEdit way of doing things.

Argmaster replied:
It also doesn't make sense to separate clearing pos1 from pos2 since you would use `/deselect` only when you want to get rid of selection entirely, there is no reason to remove just one of them, since you can always move them, only problem is that for now they permanently leave a wireframe in your world. This should change with selection wand, but now it is what it is.

Argmaster commented on src/server/command/worldedit/showpos2.zig:
```diff
@@ -0,0 +1,19 @@
+const User = main.server.User;
+
+pub const description = "Show previously selected 2nd position coordinates.";
+pub const usage = "/showpos2";
```
Flat command line interfaces are easier to use than nested ones. Since they are visualized I think we don't need them anymore.

IntegratedQuantum commented on src/server/command/worldedit/pos1.zig:
```diff
@@ -0,0 +1,22 @@
+const User = main.server.User;
+const Vec3i = main.vec.Vec3i;
+
+pub const description = "Select the player's position as position 1.";
```
```suggestion
pub const description = "Select the player position as position 1.";
```

IntegratedQuantum commented on src/server/command/worldedit/pos2.zig:
```diff
@@ -0,0 +1,22 @@
+const User = main.server.User;
+const Vec3i = main.vec.Vec3i;
+
+pub const description = "Select the player's position as position 2.";
```
```suggestion
pub const description = "Select the player position as position 2.";
```

IntegratedQuantum commented on src/server/command/worldedit/pos2.zig:
```diff
@@ -0,0 +1,22 @@
+	source.worldEditData.selectionPosition2 = pos;
+	main.network.Protocols.genericUpdate.sendWorldEditPos(source.conn, .selectedPos2, pos);
+
+	source.sendMessage("Position 1: {}", .{pos});
```
```suggestion
	source.sendMessage("Position 2: {}", .{pos});
```

IntegratedQuantum reviewed (changes requested):
I have tested this and found one small issue with blocks that require support: #1217
However, please do not add a fix to this PR, since it is non-trivial to fix and I would like to get this merged since it's blocking structures.

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,248 @@
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const User = main.server.User;
+
+pub const GameIdToBlueprintIdMapType = std.AutoHashMap(u16, u16);
```
```suggestion
const GameIdToBlueprintIdMapType = std.AutoHashMap(u16, u16);
```

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,248 @@
+		const compression = try compressedReader.readEnum(BlueprintCompression);
+		const blockPaletteSizeBytes = try compressedReader.readInt(u32);
+		const paletteBlockCount = try compressedReader.readInt(u16);
+		const width: u32 = try compressedReader.readInt(u16);
+		const depth: u32 = try compressedReader.readInt(u16);
+		const height: u32 = try compressedReader.readInt(u16);
```
The explicit widening of the type doesn't make any here in my opinion.
```suggestion
		const width = try compressedReader.readInt(u16);
		const depth = try compressedReader.readInt(u16);
		const height = try compressedReader.readInt(u16);
```

IntegratedQuantum commented on src/server/command/worldedit/copy.zig:
```diff
@@ -0,0 +1,44 @@
+
+					source.sendMessage("Copied selection to clipboard.", .{});
+				},
+				.failure => {
+					const e = result.failure;
```
You can also use a capture here:
```suggestion
				.failure => |e| {
```

IntegratedQuantum commented on src/server/command/worldedit/copy.zig:
```diff
@@ -0,0 +1,44 @@
+		return;
+	}
+
+	if(source.worldEditData.selectionPosition1) |pos1| {
```
It's generally more readable to use early returns instead of nested if elses:
```zig
const pos1 = source.worldEditData.selectionPosition1 orelse {
	return source.sendMessage("#ff0000Position 1 isn't set", .{});
};
const pos2 = source.worldEditData.selectionPosition2 orelse {
	return source.sendMessage("#ff0000Position 2 isn't set", .{});
};
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,201 @@
+	}
+	source.worldEditData.clipboard = Blueprint.load(main.globalAllocator, storedBlueprint) catch |err| {
+		std.log.warn("Failed to load blueprint file '{s}' ({s})", .{fileName, @errorName(err)});
+		source.sendMessage("#ff0000Failed to load blueprint file '{s}' ({s})", .{fileName, @errorName(err)});
```
Since you do this everywhere now, how about a helper function to avoid duplicating the messages everywhere?

IntegratedQuantum replied:
this is not resolved

Argmaster commented:
Btw extracting blueprint file io to struct made the code longer by about 20 lines 😝

IntegratedQuantum commented:
> Btw extracting blueprint file io to struct made the code longer by about 20 lines 😝

Why did you do it then?

IntegratedQuantum commented on src/blueprint.zig:
```diff
@@ -0,0 +1,248 @@
+	}
+	fn storeBlockPalette(allocator: NeverFailingAllocator, map: GameIdToBlueprintIdMapType, writer: *BinaryWriter) usize {
+		var blockPalette = allocator.alloc([]const u8, map.count());
+		defer allocator.free(blockPalette);
```
Local allocations should always use the stack allocator.

Argmaster commented:
> Why did you do it then?

Cause this seemed to be a good abstraction. I have reverted it tho, and moved just the opening into separate function, it is still +1 line of diff.

IntegratedQuantum commented:
How about just a helper function for the two output calls:
```
fn sendWarningAndLog(comptime fmt: []const u8, args: anytype) void {
    std.log.warn(fmt, args);
    user.sendMessage("#ff0000" ++ fmt, args);
}
```

Argmaster commented:
Done.

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,198 @@
+}
+
+fn sendInfoAndLog(comptime fmt: []const u8, args: anytype, user: *User) void {
+	std.log.warn(fmt, args);
```
```suggestion
	std.log.info(fmt, args);
```

IntegratedQuantum commented on src/server/command/worldedit/blueprint.zig:
```diff
@@ -0,0 +1,198 @@
+
+	blueprintsDir.dir.deleteFile(fileName) catch |err| {
+		sendWarningAndLog("Failed to delete blueprint file '{s}' ({s})", .{fileName, @errorName(err)}, source);
+		return;
```
Not sure if you want to do this, but you could also put these in one line, since the return types (void) match:
```
return sendWarningAndLog(...);
```

IntegratedQuantum commented:
Did this finally save some lines instead of adding them?

Argmaster commented:
> Did this finally save some lines instead of adding them?

 Yes, total of 4.

IntegratedQuantum reviewed (approved):
Alright, I think we are done here.

Honestly if it weren't needed for the structure models, I don't know if I would have had the motivation to keep reviewing this. In the future, please please please try to make smaller PRs. Reviewing 650 lines just takes so much time and energy, and I'm sure that's true on your side too.
